File: src/web-app/src/lib/archived-projects.test.ts
import { expect, test } from "bun:test"
import { archivedProjects, filterByProject, projectLabel, type ArchivedLike } from "./archived-projects"

const HOME = "/home/ahmet"

function s(workdir: string, opts: Partial<ArchivedLike> = {}): ArchivedLike {
  return { workdir, ...opts }
}

// --- projectLabel ---

test("projectLabel shows parent/leaf with ellipsis for nested home paths", () => {
  expect(projectLabel("/home/ahmet/projects/kurbanhane", HOME)).toBe("…/projects/kurbanhane")
})

test("projectLabel shows ~/leaf when the project sits directly in home", () => {
  expect(projectLabel("/home/ahmet/foo", HOME)).toBe("~/foo")
})

test("projectLabel shows ~ for the home directory itself", () => {
  expect(projectLabel("/home/ahmet", HOME)).toBe("~")
})

test("projectLabel keeps parent/leaf for a shallow non-home path", () => {
  expect(projectLabel("/srv/acme", HOME)).toBe("srv/acme")
})

test("projectLabel adds ellipsis for a deep non-home path", () => {
  expect(projectLabel("/srv/www/acme", HOME)).toBe("…/www/acme")
})

test("projectLabel returns a single-segment path unchanged", () => {
  expect(projectLabel("/acme", HOME)).toBe("/acme")
})

// --- archivedProjects ---

test("archivedProjects dedupes by project and counts sessions", () => {
  const result = archivedProjects([
    s("/home/ahmet/projects/foo", { killed_at: "2026-06-01T00:00:00Z" }),
    s("/home/ahmet/projects/foo", { killed_at: "2026-06-02T00:00:00Z" }),
  ], HOME)
  expect(result).toEqual([{ key: "/home/ahmet/projects/foo", label: "…/projects/foo", count: 2 }])
})

test("archivedProjects groups a worktree session under its repo_root", () => {
  const result = archivedProjects([
    s("/home/ahmet/.mux/worktrees/x/abc", { repo_root: "/home/ahmet/projects/foo", killed_at: "2026-06-01T00:00:00Z" }),
    s("/home/ahmet/projects/foo", { killed_at: "2026-06-02T00:00:00Z" }),
  ], HOME)
  expect(result).toEqual([{ key: "/home/ahmet/projects/foo", label: "…/projects/foo", count: 2 }])
})

test("archivedProjects orders most-recently-archived first", () => {
  const result = archivedProjects([
    s("/home/ahmet/projects/old", { killed_at: "2026-06-01T00:00:00Z" }),
    s("/home/ahmet/projects/new", { killed_at: "2026-06-10T00:00:00Z" }),
  ], HOME)
  expect(result.map((p) => p.label)).toEqual(["…/projects/new", "…/projects/old"])
})

test("archivedProjects returns [] for no sessions", () => {
  expect(archivedProjects([], HOME)).toEqual([])
})

// --- filterByProject ---

test("filterByProject returns only sessions in the given project", () => {
  const sessions = [s("/home/ahmet/projects/foo"), s("/home/ahmet/projects/bar")]
  expect(filterByProject(sessions, "/home/ahmet/projects/foo", HOME)).toEqual([
    s("/home/ahmet/projects/foo"),
  ])
})

test("filterByProject matches a worktree session by repo_root", () => {
  const wt = s("/home/ahmet/.mux/worktrees/x/abc", { repo_root: "/home/ahmet/projects/foo" })
  expect(filterByProject([wt], "/home/ahmet/projects/foo", HOME)).toEqual([wt])
})

test("filterByProject returns all sessions when key is null", () => {
  const sessions = [s("/home/ahmet/projects/foo"), s("/home/ahmet/projects/bar")]
  expect(filterByProject(sessions, null, HOME)).toEqual(sessions)
})

File: apps/shared/src/commonTest/kotlin/dev/supermux/session/ArchivedProjectsTest.kt
package dev.supermux.session

import dev.supermux.net.ArchivedDto
import kotlin.test.Test
import kotlin.test.assertEquals

class ArchivedProjectsTest {
    private val HOME = "/home/ahmet"

    private fun a(workdir: String, repo: String? = null, killed: String? = null) =
        ArchivedDto(id = "x", name = "n", workdir = workdir, agent = "claude", killed_at = killed, repo_root = repo)

    @Test
    fun dedupes_and_counts() {
        assertEquals(
            listOf(ArchivedProject("/home/ahmet/projects/foo", "…/projects/foo", 2)),
            archivedProjects(
                listOf(
                    a("/home/ahmet/projects/foo", killed = "2026-06-01T00:00:00Z"),
                    a("/home/ahmet/projects/foo", killed = "2026-06-02T00:00:00Z"),
                ),
                HOME,
            ),
        )
    }

    @Test
    fun groups_worktree_under_repo_root() {
        assertEquals(
            listOf(ArchivedProject("/home/ahmet/projects/foo", "…/projects/foo", 2)),
            archivedProjects(
                listOf(
                    a("/home/ahmet/.mux/worktrees/x/abc", repo = "/home/ahmet/projects/foo", killed = "2026-06-01T00:00:00Z"),
                    a("/home/ahmet/projects/foo", killed = "2026-06-02T00:00:00Z"),
                ),
                HOME,
            ),
        )
    }

    @Test
    fun orders_most_recent_first() {
        assertEquals(
            listOf("…/projects/new", "…/projects/old"),
            archivedProjects(
                listOf(
                    a("/home/ahmet/projects/old", killed = "2026-06-01T00:00:00Z"),
                    a("/home/ahmet/projects/new", killed = "2026-06-10T00:00:00Z"),
                ),
                HOME,
            ).map { it.label },
        )
    }

    @Test
    fun empty_input_yields_empty() {
        assertEquals(emptyList<ArchivedProject>(), archivedProjects(emptyList(), HOME))
    }

    @Test
    fun label_shortens_non_home_path() {
        assertEquals(
            listOf("…/www/acme"),
            archivedProjects(listOf(a("/srv/www/acme", killed = "2026-06-01T00:00:00Z")), HOME).map { it.label },
        )
    }

    @Test
    fun ties_break_alphabetically_by_label() {
        assertEquals(
            listOf("…/projects/alpha", "…/projects/beta"),
            archivedProjects(
                listOf(
                    a("/home/ahmet/projects/beta", killed = "2026-06-05T00:00:00Z"),
                    a("/home/ahmet/projects/alpha", killed = "2026-06-05T00:00:00Z"),
                ),
                HOME,
            ).map { it.label },
        )
    }

    @Test
    fun null_killed_at_sorts_last() {
        assertEquals(
            listOf("…/projects/recent", "…/projects/null-killed"),
            archivedProjects(
                listOf(
                    a("/home/ahmet/projects/null-killed", killed = null),
                    a("/home/ahmet/projects/recent", killed = "2026-06-10T00:00:00Z"),
                ),
                HOME,
            ).map { it.label },
        )
    }

    @Test
    fun filter_matches_by_key() {
        val sessions = listOf(a("/home/ahmet/projects/foo"), a("/home/ahmet/projects/bar"))
        assertEquals(
            listOf(a("/home/ahmet/projects/foo")),
            filterArchivedByProject(sessions, "/home/ahmet/projects/foo"),
        )
    }

    @Test
    fun filter_matches_worktree_by_repo_root() {
        val wt = a("/home/ahmet/.mux/worktrees/x/abc", repo = "/home/ahmet/projects/foo")
        assertEquals(listOf(wt), filterArchivedByProject(listOf(wt), "/home/ahmet/projects/foo"))
    }

    @Test
    fun filter_null_returns_all() {
        val sessions = listOf(a("/home/ahmet/projects/foo"), a("/home/ahmet/projects/bar"))
        assertEquals(sessions, filterArchivedByProject(sessions, null))
    }
}
